polkadot/xcm/xcm-executor/src/traits/controller.rs

```diff
@@ -0,0 +1,206 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! A set of traits that define how a pallet interface with XCM.
+
+use crate::{
+	traits::{QueryHandler, QueryResponseStatus},
+	InteriorMultiLocation,
+	Junctions::Here,
+	MultiLocation, Xcm, Outcome, XcmError
+};
+use frame_support::{pallet_prelude::*, parameter_types};
+use sp_weights::Weight;
+use xcm::{v3::XcmHash, VersionedMultiLocation, VersionedXcm};
+
+/// Umbrella trait for all Controller traits.
+pub trait Controller<Origin, RuntimeCall, Timeout>:
+	ExecuteController<Origin, RuntimeCall> + SendController<Origin> + QueryController<Origin, Timeout>
+{
+}
+
+impl<T, Origin, RuntimeCall, Timeout> Controller<Origin, RuntimeCall, Timeout> for T where
+	T: ExecuteController<Origin, RuntimeCall>
+		+ SendController<Origin>
+		+ QueryController<Origin, Timeout>
+{
+}
+
+/// Weight functions needed for [`ExecuteController`].
+pub trait ExecuteControllerWeightInfo {
+	/// Weight for [`ExecuteController::execute`]
+	fn execute() -> Weight;
+}
+
+/// Execute an XCM locally, for a given origin.
+pub trait ExecuteController<Origin, RuntimeCall> {
+	/// Weight information for ExecuteController functions.
+	type WeightInfo: ExecuteControllerWeightInfo;
+
+	/// Attempt to execute an XCM locally, and return the outcome.
+	///
+	/// # Parameters
+	///
+	/// - `origin`: the origin of the call.
+	/// - `message`: the XCM program to be executed.
+	/// - `max_weight`: the maximum weight that can be consumed by the execution.
+	fn execute(
+		origin: Origin,
+		message: Box<VersionedXcm<RuntimeCall>>,
+		max_weight: Weight,
+	) -> Result<Outcome, DispatchError>;
+}
+
+/// Weight functions needed for [`SendController`].
+pub trait SendControllerWeightInfo {
+	/// Weight for [`SendController::send`]
+	fn send() -> Weight;
+}
+
+/// Send an XCM from a given origin.
+pub trait SendController<Origin> {
```

Comment:
We also have the `SendXcm` trait with is meant to be a low-level message delivery trait. We can rename this to `SendVersionedXcm` or `SendXcmWrapper`, but maybe just keeping the same name but adding appropriate documentation will be enough.

Reply:
yeah I would reply same thing as above https://github.com/paritytech/polkadot-sdk/pull/2086#discussion_r1377754953

I am fine replacing  Controller with Wrapper if we think it's a better name